src/components/tf.js

```diff
@@ -0,0 +1,438 @@
+    });
+  }
+
+  nextBatch(batchSize, data, index) {
+    const batchImagesArray = new Float32Array(batchSize * IMAGE_SIZE);
+    const batchLabelsArray = new Uint8Array(batchSize * NUM_CLASSES);
+
+    for (let i = 0; i < batchSize; i++) {
+      const idx = index();
+
+      const image =
+          data[0].slice(idx * IMAGE_SIZE, idx * IMAGE_SIZE + IMAGE_SIZE);
+      batchImagesArray.set(image, i * IMAGE_SIZE);
+
+      const label =
+          data[1].slice(idx * NUM_CLASSES, idx * NUM_CLASSES + NUM_CLASSES);
+      batchLabelsArray.set(label, i * NUM_CLASSES);
+    }
+
+    const xs = tf.tensor2d(batchImagesArray, [batchSize, IMAGE_SIZE]);
+    const labels = tf.tensor2d(batchLabelsArray, [batchSize, NUM_CLASSES]);
+
+    return {xs, labels};
+  }
+}
+
+export function getModel() {
+  console.log("getModel ..");
+  model = tf.sequential();
+
+  model.add(tf.layers.conv2d({
+    inputShape: [28, 28, 1],
+    kernelSize: 5,
+    filters: 8,
+    strides: 1,
+    activation: 'relu',
+    kernelInitializer: 'varianceScaling'
+  }));
+
+  model.add(tf.layers.maxPooling2d({poolSize: [2, 2], strides: [2, 2]}));
+  model.add(tf.layers.conv2d({
+    kernelSize: 5,
+    filters: 16,
+    strides: 1,
+    activation: 'relu',
+    kernelInitializer: 'varianceScaling'
+  }));
+
+  model.add(tf.layers.maxPooling2d({poolSize: [2, 2], strides: [2, 2]}));
+  model.add(tf.layers.flatten());
+
+  model.add(tf.layers.dense({
+    units: 10,
+    kernelInitializer: 'varianceScaling',
+    activation: 'softmax'
+  }));
+
+  const LEARNING_RATE = 0.15;
+  const optimizer = tf.train.sgd(LEARNING_RATE);
+
+  model.compile({
+    optimizer: optimizer,
+    loss: 'categoricalCrossentropy',
+    metrics: ['accuracy'],
+  });
+
+  return model;
+}
+
+export async function loadData() {
+  data = new MnistData();
+  await data.load();
+  return data;
+}
+
+async function initData() {
+  //data = await loadData();
+  data = new MnistData();
+  await data.load();
+  
+  const surface = tfvis.visor().surface({
+    name: 'Surface',
+    tab: 'Input Data'
+  });
+
+  const drawArea = surface.drawArea; // Get the examples
+
+  const examples = data.nextTestBatch(20);
+  const numExamples = examples.xs.shape[0];
+
+  for (let i = 0; i < numExamples; i++) {
+    const imageTensor = tf.tidy(() => {
+      return examples.xs.slice([i, 0], [1, examples.xs.shape[1]]).reshape([28, 28, 1]);
+    }); // Create a canvas element to render each example
+
+    const canvas = document.createElement('canvas');
+    canvas.width = 28;
+    canvas.height = 28;
+    canvas.style = 'margin: 4px;';
+    tf.browser.toPixels(imageTensor, canvas);
+    drawArea.appendChild(canvas);
+    imageTensor.dispose();
+  }
+  console.log("loadData done");
+}
+
+function initModel() {
+  console.log("initModel ...");
+  window.model = getModel();
+}
+
+// doesn't work (fix later)
+function example_button() {
+  document.querySelector('#show-examples').addEventListener('click', async (e) => {
```

Comment:
i don't know if DOM manipulation is in line with React expectations, try to read more about it? doesn't necessarily have to be fixed in this PR